FEM-2211: Track selection improvements

noamtamim opened the pull request:
New feature: API for selection of 
- Video codec, bitrate, resolution
- Audio languages
- Text languages

Selection is done by instantiating and configuring a `DTGSelectionOptions` object and passing it to the new `DTGContentManager.loadItemMetadata(id: String, options: DTGSelectionOptions?)` function. This new function is used **instead** of `loadItemMetadata(id: String, preferredVideoBitrate: Int?)`.

noamtamim committed:
Initial work on track selection

noamtamim committed:
Add docs for the new track selection API

noamtamim committed:
Improve API

noamtamim committed:
Rearrange the DTGSelectionSettings class

noamtamim committed:
Add allowInefficientCodecs to selection settings

noamtamim committed:
Fix the API and initial implementation

For now, the only supported option is VideoBitrate.avc1().

noamtamim committed:
Start implementation

noamtamim committed:
Updated GCDWebServer and Realm

noamtamim committed:
Don't use the default XCGLogger

noamtamim committed:
Use published playkit in the sample app

noamtamim committed:
Implement **video** stream selection

Taking into account:
Codec (avc1 or hevc)
Preferred video height and width
Preferred bitrate
Preferred codec

noamtamim committed:
Travis: use xcode 10.1

noamtamim committed:
Log: show tag, don't show timestamp

noamtamim committed:
Code cleanup

noamtamim committed:
More cleanup

noamtamim committed:
More cleanup...

noamtamim committed:
Refactor and handle audio groups etc

noamtamim committed:
Refactor and don't download filtered audio and text

noamtamim committed:
Fix copy-paste mistake

noamtamim committed:
Copy M3U8Kit source to DTG

Need to extend it.

noamtamim committed:
Remove `import M3U8Kit`

noamtamim committed:
More fine tuning

noamtamim committed:
Use another approach to build the local master

noamtamim committed:
Back to using the M3U8Kit pod

noamtamim committed:
Add CODECS to local master and fix nil crash

noamtamim committed:
Remove the now unused code

noamtamim committed:
Use published PlayKit

noamtamim committed:
Add basic unit testing

noamtamim committed:
Print playable and filtered video streams in debug

noamtamim committed:
Remove Quick and Nimble

noamtamim committed:
More tests

noamtamim committed:
Sort streams by bandwidth

noamtamim committed:
Annotate the setters of DTGSelectionOptions with `@discardableResult`

noamtamim committed:
More tests

noamtamim committed:
Move the selection API to another file

noamtamim committed:
Fix bug in local master line

noamtamim committed:
New download/play tests

noamtamim committed:
Test fixes

noamtamim committed:
Fix download test code

noamtamim committed:
Move API+Selection.java to the correct place

noamtamim committed:
ViewController.items is var

noamtamim committed:
Test and demo enhancements

noamtamim committed:
Put the json in resources

noamtamim committed:
Put some items in items.json and make travis test

noamtamim committed:
Travis: upload artifacts

noamtamim committed:
Try Xcode 10.2 beta

noamtamim committed:
travis test: don't specify ios version

noamtamim committed:
Don't try to prepare/play on travis

noamtamim committed:
Sample: show item info when tapping "progress"

noamtamim committed:
Another attempt at not playing in travis

noamtamim committed:
Update travis-build.sh

noamtamim committed:
Test download and play on xcode 9.4 (pre-mohave)

noamtamim committed:
Merge branch 'FEM-2211' of ssh://github.com/kaltura/playkit-ios-dtg into FEM-2211

noamtamim committed:
Back to Xcode 10.1

noamtamim committed:
Travis: send build info to jenkins

noamtamim committed:
Add travis build url as jenkins cause

noamtamim committed:
Don't abort the build if it failed (but report it)

noamtamim committed:
Use xcpretty-travis-formatter on travis

noamtamim committed:
Actually enable the travis formatter

noamtamim committed:
xcpretty-travis-formatter doesn't really work

noamtamim committed:
More detailed download fulfill

noamtamim committed:
Revert "More detailed download fulfill"

This reverts commit e3764c4d7dcef3ae97c1fb6fedec69fddd4bb6f2.

noamtamim committed:
More detailed download completion check

noamtamim committed:
Try delaying the item state notification

noamtamim committed:
Test 8 times the basic content

noamtamim committed:
Add tests with more options, fix the travis WA

x-NR-x commented on Sources/API+Selection.swift:
```diff
@@ -0,0 +1,202 @@
+    /// bitrate is not defined.
+    public var videoBitrates: [VideoBitrate]? = nil {
+        didSet {
+            videoCodecs = nil
```
If it was set with nil, it will also set the videoCodecs to nil, was this your intension?

noamtamim replied:
Fixed

x-NR-x commented on Sources/API+Selection.swift:
````diff
@@ -0,0 +1,202 @@
+    /// ```
+    public var textLanguages: [String]? = nil {
+        didSet {
+            if textLanguages != nil {
````
Need to add a check that it in not an empty array.

noamtamim replied:
Added to doc:

> An empty array means not downloading text tracks at all.

x-NR-x commented on Sources/API+Selection.swift:
````diff
@@ -0,0 +1,202 @@
+    /// ```
+    public var audioLanguages: [String]? = nil {
+        didSet {
+            if audioLanguages != nil {
````
Need to add a check that it in not an empty array.

noamtamim replied:
Same

noamtamim committed:
Fix FEM-2444 (FairPlay)

The EXT-X-SESSION-KEY tag for FairPlay was not copied to the local master.

noamtamim committed:
Fix FEM-2443

Old-style HLS assets contain audio-only tracks. Without a minimum bitrate, DTG would select them.

The fix is setting a default minimum bitrate of 180kb (AVC1) or 120kb (HEVC).

noamtamim committed:
Doc clarification, code review comments
